pack: Fix read Stack-buffer-overflow READ in msgpack_sbuffer_write

sashashura opened the pull request:
This PR fixes Stack-buffer-overflow in msgpack_sbuffer_write revealed by fuzzing:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=45208

The root cause is that snprintf call in flb_pack_msgpack_to_json_format returns the number of characters that would have been written if the buffer had been sufficiently large.
It leads to reading out of the boundaries of the buffer later.

sashashura commented:
Hopefully now everything is correct. It truncates the time, but it is as it was, I was afraid to break anything. Let me know if you think it is better to make the buffer bigger.

sashashura commented:
Anything I need to do?

sashashura commented:
@nokute78?

nokute78 reviewed (changes requested):
I wrote below comment at previous PR.
> If len is larger than `sizeof(time_formatted) - 1 -s`, we need to resize the buffer or return error.

Currently, your patch prevent memory leak and it is good.
However output may be truncated silently.
I think we need to notify users or prevent to truncate.

sashashura commented:
Sorry, missed that request.

It is not about memory leaks, but about reading stack buffer beyond bounds.
The `char time_formatted[32];` is first filled with 19 characters like `1970-01-01 00:00:00`.
Then `snprintf` returns 18 - how much it would need to print microseconds. It is added to 19: `s += len;` and gives 37 which is 5 bytes bigger than the size of the `time_formatted` buffer. So the `msgpack_pack_str_body(&tmp_pck, time_formatted, s);` call results in reading beyond the buffer bounds.

Reading the documentation I think you are doing minus one for no reason:

>Provided that the result string, including the terminating null byte, does not exceed max bytes, strftime() returns the number of bytes (excluding the terminating null byte) placed in the arrays.

and

>The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.

So I have removed that and increased the buffer to fit microseconds. It should always fit now, but I also added the `return` if it would be truncated.

sashashura commented:
@nokute78 Ready for review.

nokute78 commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
+                    if (json_format == FLB_PACK_JSON_FORMAT_LINES ||
+                        json_format == FLB_PACK_JSON_FORMAT_STREAM) {
+                            flb_sds_destroy(out_buf);
+                        }
```
Could you check indentation level ?

sashashura replied:
I thought it will be more readable. Anyway done.

nokute78 replied:
Oops, I meant like below.
Could you fix it ?

```c
                    if (json_format == FLB_PACK_JSON_FORMAT_LINES ||
                        json_format == FLB_PACK_JSON_FORMAT_STREAM) {
                        flb_sds_destroy(out_buf);
                    }
                    msgpack_sbuffer_destroy(&tmp_sbuf);
```

sashashura replied:
Now I see it! Done.

nokute78 reviewed (changes requested):
I added a request.

> Then snprintf returns 18 - how much it would need to print microseconds.

How does it return 18 ?

sashashura commented:
> > Then snprintf returns 18 - how much it would need to print microseconds.
> 
> How does it return 18 ?

`tms.tm.tv_sec` is -100000000. Then `(uint64_t) tms.tm.tv_nsec / 1000` gives 18446744073609551 which requires 18 bytes to format with '\0'.

While you are here... May I ask why https://github.com/fluent/fluent-bit/pull/5727 is not merged yet?

nokute78 reviewed (commented):
> tms.tm.tv_sec is -100000000. Then (uint64_t) tms.tm.tv_nsec / 1000 gives 18446744073609551 which requires 18 bytes to format with '\0'.

OK. I understood.

>  While you are here... May I ask why https://github.com/fluent/fluent-bit/pull/5727 is not merged yet?

I don't have a permission to merge `flb_ra_key.c` and `flb_pack.c`.
Please be waiting to merge by @edsiper .

nokute78 reviewed (approved):
@sashashura  Thank you! LGTM.

@edsiper Could you merge this ?

sashashura commented:
Folks, what is going on?

ptsneves commented:
Ping

DavidKorczynski commented:
Regarding the truncation, something similar happened previously: https://github.com/fluent/fluent-bit/pull/2798

patrick-stephens commented:
@sashashura can you add the Valgrind output as per the template?

leonardo-albertovich commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
                 len = snprintf(time_formatted + s,
-                               sizeof(time_formatted) - 1 - s,
+                               max_len,
                                ".%06" PRIu64,
```
I think there's an error in the worst case scenario and this `snprintf` could use up to 21 characters (leading dot + 20 characters which is the longest number a `uint64_t` can hold). Additionally, the `snprintf` in line 954 could use up to 22 characters (leading dot, uint64 and trailing T).

So considering we're not so constrained when it comes to memory, I'd make `time_formatted` at least 40 characters long.

leonardo-albertovich replied:
Yeah, I know it shouldn't but considering the context obsessing over a few stack bytes makes no sense to me. 
However, it's not a huge concern so it can stay like that.

leonardo-albertovich commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
                              FLB_PACK_JSON_DATE_JAVA_SQL_TIMESTAMP_FMT, &tm);
 
+                max_len = sizeof(time_formatted) - s;
                 len = snprintf(time_formatted + s,
```
Could you change the way in which you get the buffer pointer from `time_formatted + s` to `&time_formatted[s]`?

leonardo-albertovich commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
             /* Format the time, use microsecond precision not nanoseconds */
                 gmtime_r(&tms.tm.tv_sec, &tm);
-                s = strftime(time_formatted, sizeof(time_formatted) - 1,
+                s = strftime(time_formatted, sizeof(time_formatted),
```
Since there are no real reasons for `strftime` to return 0 with the provided arguments could you add a line to verify the result and abort on error?

leonardo-albertovich commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
+                s = strftime(time_formatted, sizeof(time_formatted),
                              FLB_PACK_JSON_DATE_JAVA_SQL_TIMESTAMP_FMT, &tm);
 
+                max_len = sizeof(time_formatted) - s;
```
I would prefer it if instead of setting `max_len` here you set `max_len` to `sizeof(time_formatted)` before line 927, used the variable as the argument in the `strftime` call and only subtract `s` from it in line 930.

leonardo-albertovich commented on src/flb_pack.c:
```diff
@@ -923,13 +924,23 @@ flb_sds_t flb_pack_msgpack_to_json_format(const char *data, uint64_t bytes,
                                ".%06" PRIu64,
                                (uint64_t) tms.tm.tv_nsec / 1000);
+                if (len >= max_len) {
+                    if (json_format == FLB_PACK_JSON_FORMAT_LINES ||
```
I think in reality what we care about at this point is not what `json_format` holds but to ensure that `out_buf` is not `NULL`, in fact there's an error related to that in a different point in the function so, let's keep it as simple as possible.

sashashura replied:
I have completely removed the check because it is safe to call the cleanup function with NULL.

leonardo-albertovich reviewed (changes requested):
I would like to merge this as soon as possible but before doing so I'd need these changes performed. 

Additionally, the same pattern is repeated in line 948 so it would make sense to fix it there too and since both blocks are the exact same thing except for the `strftime` format string and the microsecond trailer it would make sense to extract this logic into a smaller static function that gets these and the as arguments. 

It could either just format the date to a buffer or format it and append it to the msgpack buffer ( `msgpack_pack_formatted_datetime` sounds like an appropriate name.).

To be honest, fixing the second block or extracting the logic is not a hard requirement but it would make sense  since we're touching this.

Additionally, could you write a message using `flb_debug` in the place where you abort the process?

Bonus : it's unrelated to your PR but the `flb_sds_destroy` call in line 879 should be removed since there's no way for `out_buf` to be initialized if the code takes that path and it doesn't even check it.

sashashura commented:
Done. @leonardo-albertovich please take a look.

leonardo-albertovich commented:
Looks good to me, thanks for extracting the functionality and fixing the remaining block in the process!
